ichnaea/log.py

```diff
@@ -179,54 +179,78 @@ def configure_stats():
+        structlog.threadlocal.clear_threadlocal()
+        structlog.threadlocal.bind_threadlocal(
+            http_method=request.method, http_path=request.path
+        )
+        # Skip detailed logging and capturing for static assets, either in
+        # /static or paths like /robots.txt
+        full_logs = not (
+            request.path in registry.skip_logging or request.path.startswith("/static")
+        )
 
-        def counter_send(status_code):
-            METRICS.incr("request", tags=statsd_tags + ["status:%s" % status_code])
+        def record_response(status_code):
+            """Time request, (maybe) emit metrics, and (maybe) log this request.
+
+            For static assets, metrics are skipped, and logs are skipped unless
+            we're in the development environment.
+            """
+            duration = time.time() - start
+
+            if full_logs:
+                # Emit a request.timing and a request metric
+                duration_ms = int(round(duration * 1000))
+                statsd_tags = [
+                    # Convert a URI to a statsd acceptable metric name
+                    "path:%s"
+                    % request.path.replace("/", ".").lstrip(".").replace("@", "-"),
+                    "method:%s" % request.method.lower(),
+                ]
+                METRICS.timing("request.timing", duration_ms, tags=statsd_tags)
+                METRICS.incr("request", tags=statsd_tags + ["status:%s" % status_code])
+
+            if local_dev_env or full_logs:
+                # Emit a canonical-log-line
+                duration_s = round(duration, 3)
+                logger = structlog.get_logger("canonical-log-line")
+                logger.info(
+                    f"{request.method} {request.path} - {status_code}",
+                    http_status=status_code,
+                    duration_s=duration_s,
+                )
 
         try:
             response = handler(request)
-            timer_send()
-            counter_send(response.status_code)
+            record_response(response.status_code)
             return response
         except (BaseClientError, HTTPRedirection) as exc:
-            # don't capture exceptions
-            timer_send()
-            counter_send(exc.status_code)
+            # BaseClientError: 4xx error raise by Ichnaea API, other Ichnaea code
+            # HTTPRedirection: 3xx redirect from Pyramid
+            # Log, but do not send these exceptions to Sentry
+            record_response(exc.status_code)
             raise
         except HTTPClientError:
-            # ignore general client side errors
+            # HTTPClientError: 4xx error from Pyramid
+            # Do not log or send to Sentry
+            raise
+        except HTTPException as exc:
+            # HTTPException: Remaining 5xx (or maybe 2xx) errors from Pyramid
+            # Log, and maybe send to Sentry
+            record_response(exc.status_code)
+            if full_logs:
```

Comment:
I'm not sure I understand this. Why `full_logs` and not `local_dev_env`? Why wouldn't you want to send the exception here?